test/test_autocast.py

```diff
@@ -0,0 +1,452 @@
+      output = output_method = None
+
+      # Try module.* variant, if requested:
+      if module is not None and hasattr(module, op):
+        output = getattr(module, op)(*args, **add_kwargs)
+        if isinstance(output, torch.Tensor):
+          self.assertTrue(
+              out_type == output.dtype,
+              "autocast for torch.{} produced {}, should produce {}".format(
+                  op, output.dtype, out_type))
+
+      # Try Tensor.* variant:
+      if hasattr(torch.Tensor, op):
+        output_method = getattr(args[0], op)(*args[1:], **add_kwargs)
+        if isinstance(output_method, torch.Tensor):
+          self.assertTrue(
+              out_type == output_method.dtype,
+              "autocast for torch.{} produced {}, should produce torch.{}".
+              format(op, output_method.dtype, out_type))
+
+      self.assertTrue((output is not None) or (
+          output_method is not None
+      ), "{} not found as an attribute on either Tensor or the requested module {}"
+                      .format(op, module))
+
+      # Accounts for ops that return Tensors, iterables, and other non-Tensors.
+      # For example, lstm_cell returns a tuple and equal returns bool.
+      def compare(first, second):
+        if isinstance(first, torch.Tensor):
+          return torch.equal(first, second)
+        elif isinstance(first, collections.abc.Iterable):
+          return all(compare(f, s) for f, s in zip(first, second))
+        else:
+          return first == second
+
+      # If both torch.* and Tensor.* variants were found, check outputs are identical
+      if (output is not None) and (output_method is not None):
+        self.assertTrue(type(output) == type(output_method))
+        comparison = compare(output, output_method)
+        self.assertTrue(
+            comparison,
+            "torch.{0} result did not match Tensor.{0} result".format(op))
+
+      # Compare numerics to Python-side "autocasting" that (we expect) does the same thing
+      # as the C++-side autocasting, and should be bitwise accurate.
+      output_to_compare = output if output is not None else output_method
+      with autocast(enabled=False):
+        self.assertFalse(torch.is_autocast_enabled())
+
+        if module is not None and hasattr(module, op):
+          control = getattr(module, op)(*cast(args, run_as_type), **add_kwargs)
+        else:
+          control = getattr(args[0].to(run_as_type),
+                            op)(*cast(args[1:], run_as_type), **add_kwargs)
+        self.assertTrue(type(output_to_compare) == type(control))
+        comparison = compare(output_to_compare, control)
+        self.assertTrue(comparison,
+                        "torch.{} result did not match control".format(op))
+      self.assertTrue(torch.is_autocast_enabled())
+    self.assertFalse(torch.is_autocast_enabled())
+
+
+class TestAutocast(TestAutocastBase):
+
+  def test_autocast_torch_fp16(self):
```

Comment:
It would probably nice if xla and cuda shared more of the autocast testing infra :) Since the xla and cuda autocast functionality is pretty similar (minus a few cases that xla doesn't handle), do you think that would be reasonable?

I need to look more closely at how the existing autocast tests are structured - maybe in the same way that `pytorch_test_base.py` in pt/xla specifies which tests to skip, we could put the info about which xla autocast tests to skip there? That would only make sense if pt/xla's usage of autocast is meant to a subset of what we have in core though.

Reply:
I think at least for now, `pt/xla autocast` is more of a `xla::gpu` autocast. The challenge part of adopting gpu `test_autocast`
1. it is in `test_cuda.py` and has quite a lot of flag like `@unittest.skipIf(not TEST_CUDNN, 'CUDNN not available')` which I am not sure how well it work with pt/xla env. (When building with pt/xla, we didn't enable CUDA build for pytorch)
2. It has `torch.cuda.amp.autocast()` hardcoded. pytorch xla's amp is an alias of `torch.cuda.amp.autocast()` but running pt/xla test against a cuda flag seems a bit weird
3. For the most of the python test, if we don't have the lowering, test will fallback to cpu and it will still pass. For the amp test, cpu doesn't necessary can handle the f16 computation which will throw an error. In other word it is more likely we break amp tests.

Reply:
I think all 3 of those challenges have workarounds. I think the decision of generalizing the tests vs. copy-pasting them into pt/xla comes down to which situation we think is more likely:
- autocast behavior/support changes in a way that's agnostic to cuda vs. xla (in which case it'll be a pain to remember to make corresponding changes to this file)
- xla autocast support diverges from cuda autocast support (in which case it'll be easier if these two tests are separated since their tests will diverge more)

for the 3 challenges:
- `@unittest.skipIf(not TEST_CUDNN, 'CUDNN not available')` can probably turn into something like "skip if not CUDNN && device=cuda" (or a less hacky version of that)
-  add a dictionary, like `autocast_dict[device('cuda')] = 'torch.cuda.amp.autocast'; autocast_dict[device('xla')] = 'torch_xla.amp.autocast'`? Kind of hacky though
- for the fallbacks, specify which autocast tests are skipped in `xla/test/pytorch_test_base.py` and plumb that into the amp tests?

Reply:
Although I guess if there's ever a change to the autocast that breaks both cuda and xla and someone fixes the cuda tests, we'll get just get pinged for the xla autocast failures. So if you think that won't happen to often (autocast changes are infrequent) and those changes above are kinda intrusive, I'm won't die on this hill 😛 . Just calling it out, but I'm fine with landing.

Reply:
I think if we want to reuse the cuda test for `autocast `fully, we need to refactor the `test_cuda` to `test_cuda_autocast` and `test_cuda_everythingelse` since we don't want to run other tests like `test_cudnn_multiple_threads_same_device`. We also need to make some effort to make it more extensible(like no passing `torch.device('cuda:0')` directly, have a way to skip some unsupported op given an unsupported list etc).  This doesn't seems super complicated but does take some time.

For now I think I can remove pt/xla custom`AutocastTestLists` and add an unsupported list on the pt/xla end.  This will cut the line I copied from cuda to ~100 lines(mostly `TestAutocastBase` I refactored from `TestCuda`). If the cuda test is refactored in a way that is more extensible we can get rid of `TestAutocastBase` and inherit from that class directly. With this change, at least we can get better coverage when pytorch update `AutocastTestLists` for cuda. WDYT?

Reply:
SGTM!